backend/regalloc/regalloc_split_state.ml

```diff
@@ -51,6 +51,157 @@ let log_renaming_info : t -> unit =
+        let[@inline] update_occuring :
+            type a. Reg.Set.t -> a Cfg.instruction -> Reg.Set.t =
+         fun occurring instr ->
+          let occurring = Reg.add_set_array occurring instr.arg in
+          let occurring = Reg.add_set_array occurring instr.res in
+          occurring
+        in
+        let occurring =
+          DLL.fold_left block.body
+            ~init:(update_occuring occurring block.terminator)
+            ~f:update_occuring
+        in
+        { destroyed; occurring })
+      loop
+      { destroyed = Reg.Set.empty; occurring = Reg.Set.empty }
+
+  (* If a temporary is destroyed/reloaded inside a loop (typically because it is
+     live at a destruction point), but never used in the loop then the
+     destructions/definitions can be moved. *)
+  let optimize_loop cfg ~header (loop : Label.Set.t)
+      (destructions_at_end, definitions_at_beginning) =
+    if debug
+    then (
+      log "ExtractSpillsAndReloadsFromLoops.optimize_loop";
+      indent ());
+    let { destroyed; occurring } =
+      compute_loop_sets cfg loop destructions_at_end
+    in
+    let to_move = Reg.Set.diff destroyed occurring in
+    match Reg.Set.is_empty to_move with
+    | true ->
+      if debug
+      then (
+        log "no move available";
+        dedent ());
+      destructions_at_end, definitions_at_beginning
+    | false ->
+      if debug
+      then
+        Reg.Set.iter (fun reg -> log "%a can be moved" Printreg.reg reg) to_move;
+      let (destructions_at_end, definitions_at_beginning)
+            : destructions_at_end * definitions_at_beginning =
+        Label.Set.fold
+          (fun label (destructions_at_end, definitions_at_beginning) ->
+            let destructions_at_end =
+              Label.Map.update label
+                (function
+                  | None -> None
+                  | Some (kind, regs) -> Some (kind, Reg.Set.diff regs to_move))
+                destructions_at_end
+            in
+            let definitions_at_beginning =
+              Label.Map.update label
+                (function
+                  | None -> None | Some regs -> Some (Reg.Set.diff regs to_move))
+                definitions_at_beginning
+            in
+            destructions_at_end, definitions_at_beginning)
+          loop
+          (destructions_at_end, definitions_at_beginning)
+      in
+      (* Add destructions before the loop. *)
+      let all_loop_predecessors : Label.Set.t =
+        (Cfg.get_block_exn cfg header).predecessors
+      in
+      let destructions_at_end : destructions_at_end =
+        Label.Set.fold
+          (fun label acc ->
+            if debug then log "destructions now happen at %a" Label.print label;
+            Label.Map.update label
+              (function
+                | None -> Some (Destruction_on_all_paths, to_move)
+                | Some (_kind, regs) ->
+                  (* CR xclerc for xclerc: double-check the kind is correct. *)
```

Comment:
Is there a reason this can't be `Destruction_only_on_exceptional_path`?

Reply:
I just double-checked; `Destruction_only_on_exceptional_path`
is for when the normal edges but exceptional one do. So, I think
that makes the kind correct here. But it is not optimal: if the
predecessor has multiple successors, we want to have the
destruction only on the edge leading to the loop. The easiest
solution is to insert blocks where needed.

Reply:
Sorry, I missed the notification. I think it is
always safe to change `only_on_exception`
to `on_all_paths`, but it might create more
spills.

In this particular case, I think it is (very?)
unlikely we will switch from one kind to the
other because that would mean "entering"
a loop through an exceptional edge.

That being said, as alluded to in a previous
comment, this pull is suboptimal (indepen-
dently of the exception case) because it
adds a destruction point on every prede-
cessor. I will work on a patch to address
that, and -depending on its complexity-
either push it to this branch, or create a
new pull request.